### PR TITLE
Add Catalog Builder

### PR DESCRIPTION
**Pull request by erikerikson**
Add a serverless service that builds a product catalog.
This service defines a dynamodb table for materializing our last best understanding of the product catalog into.
This service defines a lambda that processes product/create events on the stream to populate the catalog.

README.md describes the service
serverless.yml defines the service
package.json brings in the schema validator
catalog.js defines the code for the lambda
catalog.spec.js defines a shell into which we can pour test implementations

Temporary: // TODO get from better source
  copy of the retail-stream-schema-ingress.json schema
  copy of the product-create-schema.json

### CODE REVIEW

**Comment by erikerikson** [comment]
TODO: separate the lambda from the base project in order to reduce the weight of the lambda package.  (did you really want mocha, et cetera in your Lambda?)

**Comment by erikerikson** [comment]
TODO done.

File: product-catalog/catalog.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        callback(`${constants.MODULE} ${constants.METHOD_PUT_PRODUCT} - no records received.`);
+      }
+    } catch (ex) {
+      console.log(`${constants.MODULE} ${constants.METHOD_PUT_PRODUCT} - exception: ${JSON.stringify(ex, null, 2)}`);
+      callback(ex);
+    }
+  },
+};
+
+console.log(`${constants.MODULE} ${constants.METHOD_PUT_PRODUCT} - ENV: ${JSON.stringify(process.env, null, 2)}`);
```
Comment (gwsii) [inline product-catalog/catalog.js]: Add a TODO: remove logging?

File: product-catalog/serverless.yml [hunk 2]
```diff
@@ -42,7 +44,22 @@ resources:
...
         RetentionInDays: 7
-    # Product Catalog
+    # Product Catalog Tables
+    ProductCategory:
+      Type: 'AWS::DynamoDB::Table'
+      Properties:
+        AttributeDefinitions:
+          - AttributeName: category
+            AttributeType: S
+        KeySchema:
```
Comment (gwsii) [inline product-catalog/serverless.yml]: Weren't we going to store the 'count' of the products in that category too? Or are you just not declaring that field?

**Comment by erikerikson** [comment]
As discussed, we will be querying the ProductCatalog table's Category index for counts.  Both if we want to display counts and also if we want to remove categories in response to soft deletes on products.  If we do this we will have to change the code in catalog.js to process the messages in series rather than in parallel as it does now in order to ensure no timing/ordering issues.

**Pull request merged**